types/indexer_block_events.go

```diff
@@ -0,0 +1,93 @@
+package types
+
+import (
+	"time"
+)
+
+type IndexerBlockEventManager struct {
+	height   uint32
+	time     time.Time
+	txHashes []string
+	// map from tx hash to list of tx events
+	txEventsMap map[string][]*IndexerTendermintEvent
+}
+
+// NewIndexerBlockEventManager returns a new IndexerBlockEventManager.
+// This should be called in BeginBlocker.
+func NewIndexerBlockEventManager(height uint32, time time.Time) *IndexerBlockEventManager {
+	return &IndexerBlockEventManager{
+		txHashes:    []string{},
+		txEventsMap: make(map[string][]*IndexerTendermintEvent),
+		height:      height,
+		time:        time,
+	}
+}
+
+// MergeEvents merges the events from another block event manager into this one if
+// the block heights are the same.
+func (eventManager *IndexerBlockEventManager) MergeEvents(mgr *IndexerBlockEventManager) {
+	if eventManager.height != mgr.height {
+		return
+	}
+	for _, txHash := range mgr.txHashes {
+		events := mgr.txEventsMap[txHash]
+		for _, event := range events {
+			eventManager.AddTxnEvent(txHash, event.Subtype, event.Data)
+		}
+	}
+}
+
+// AddTxnEvent adds a transaction event to the block event manager. If the transaction hash is not already in the
+// block event manager, it is also added.
+func (eventManager *IndexerBlockEventManager) AddTxnEvent(txHash string, subType string, data string) {
+	event := IndexerTendermintEvent{
+		Subtype: subType,
+		Data:    data,
+	}
+	if txEvents, ok := eventManager.txEventsMap[txHash]; ok {
+		eventManager.txEventsMap[txHash] = append(txEvents, &event)
+	} else {
+		eventManager.txHashes = append(eventManager.txHashes, txHash)
+		eventManager.txEventsMap[txHash] = []*IndexerTendermintEvent{&event}
+	}
+}
+
+func (eventManager *IndexerBlockEventManager) GetHeight() uint32 {
+	return eventManager.height
+}
+
+func (eventManager *IndexerBlockEventManager) GetTime() time.Time {
+	return eventManager.time
+}
+
+// ProduceBlock returns the block. It should only be called in EndBlocker. Otherwise, the block is
```

Comment:
Same comment as with the constructor, how do we enforce that this is only called in `EndBlocker`?